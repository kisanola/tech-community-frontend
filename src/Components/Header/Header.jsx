import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import './Header.scss';

export class Header extends Component {
  state = {
    menu: false,
  };

  toggleMenu = () => {
    this.setState({
      menu: !this.state.menu,
    });
  };

  renderUser = () => {
    const {
      isAuth,
      match: { path },
    } = this.props;
    const { menu } = this.state;

    if (!isAuth) {
      return (
        <div className="user-info">
          <div className="user-info__notif">
            <span className="counter">12</span>
            <i className="fas fa-bell" alt="User avatar" />
          </div>
          <div className="user-info__message">
            <span className="counter">12</span>
            <i className="fas fa-envelope" alt="User avatar" />
          </div>
          <div className="user-info__avatar" alt="User avatar" />
        </div>
      );
    }
    return (
      <div className={`collapse navbar-collapse ${menu ? 'show' : ''}`}>
        <ul className="navbar-nav mr-auto">
          <li className={`nav-link ${path === '/signup' ? 'active' : ''}`}>
            <Link className="nav-link" to="/signup">
              SIGNUP
            </Link>
          </li>
          <li className={`nav-link ${path === '/login' ? 'active' : ''}`}>
            <Link className="nav-link" to="/login">
              LOGIN
            </Link>
          </li>
        </ul>
      </div>
    );
  };

  render() {
    return (
      <nav className="navbar sticky-top navbar-expand-lg navbar-light bg-white">
        <div id="sticky-nav" className="container-fluid">
          <Link className="navbar-brand" to="/">
            Navbar
          </Link>
          <div className="nav-search">
            <input type="search" placeholder="Search.." aria-label="Search" />
            <i className="fa fa-search" />
          </div>
          {this.renderUser()}
          <button
            onClick={this.toggleMenu}
            className="navbar-toggler"
            type="button"
            data-toggle="collapse"
            aria-expanded="false"
            aria-label="Toggle navigation"
          >
            <span className="navbar-toggler-icon" />
          </button>
        </div>
      </nav>
    );
  }
}

export const mapStateToProps = ({ currentUser: { user, isAuth } }) => ({
  isAuth,
  user,
});

export default connect(mapStateToProps)(Header);
