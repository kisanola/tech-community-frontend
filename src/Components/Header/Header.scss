@import '../../styles/colors';

#sticky-nav {
  padding: 0;
  .navbar-collapse {
    flex-grow: unset;
  }
}
.nav-left {
  .navbar-toggler {
    margin-left: 0.5rem;
  }
}

.nav-search {
  background-color: $surface;
  display: flex;
  align-items: center;
  padding: 0.2rem 1rem;
  border-radius: 1.5rem;
  -webkit-border-radius: 1.5rem;
  -moz-border-radius: 1.5rem;
  -ms-border-radius: 1.5rem;
  -o-border-radius: 1.5rem;
  min-width: 400px;
  max-width: 100%;
  input {
    outline: none;
    border: 0;
    background-color: transparent;
    flex: 1;
    height: 2rem;
  }
  i {
    padding-right: 0.3rem;
  }
}

@mixin notif-counter {
  .counter {
    position: absolute;
    top: -5px;
    right: -6px;
    font-size: 8px;
    display: flex;
    justify-content: center;
    align-items: center;
    background-color: red;
    color: #fff;
    font-weight: 600;
    width: 15px;
    height: 15px;
    border-radius: 50%;
    -webkit-border-radius: 50%;
    -moz-border-radius: 50%;
    -ms-border-radius: 50%;
    -o-border-radius: 50%;
  }
}

.user-info {
  display: inline-flex;
  align-items: center;
  justify-content: flex-end;
  &__names {
    font-size: 1rem;
    margin-right: 1rem;
  }
  &__notif {
    position: relative;
    margin-right: 2rem;
    @include notif-counter();
    .fa-bell {
      font-size: 25px;
      color: gray;
    }
  }
  &__message {
    position: relative;
    margin-right: 2rem;
    @include notif-counter();
    .fa-envelope {
      font-size: 25px;
      color: gray;
    }
  }
  &__avatar {
    width: 40px;
    height: 40px;
    background-color: $surface;
    border-radius: 50%;
    -webkit-border-radius: 50%;
    -moz-border-radius: 50%;
    -ms-border-radius: 50%;
    -o-border-radius: 50%;
    box-shadow: 0 0 2px 0px rgba(0, 0, 0, 0.3);
  }
}

@media screen and (max-width: 768px) {
  #sticky-nav {
    .nav-search {
      order: 2;
      width: 100%;
      margin-top: 0.5rem;
      min-width: 100%;
    }
    .navbar-collapse {
      order: 2;
      text-align: center;
      .navbar-nav {
        .nav-link {
          cursor: pointer;
        }
      }
    }
  }
}
