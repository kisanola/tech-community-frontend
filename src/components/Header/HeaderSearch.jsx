
import {
  InstantSearch,
  SearchBox,
  connectAutoComplete,
} from 'react-instantsearch-dom';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import searchClient from '../../helpers/searchClient';
import './HeaderSearch.scss';
import { setSearchState } from '../../actions/searchActions';

const Autocomplete = ({ hits }) => (
  <ul className="search-autocomplete">
    {hits.map((hit) => (
      <li key={hit.objectID}>
        <Link to={`/${hit.resource}s/${hit.objectID}`}>
          <p>{hit.title}</p>
          <span>
            <i className="fa fa-list" />
            {hit.resource}
          </span>
        </Link>
      </li>
    ))}
  </ul>
);

const CustomAutocomplete = connectAutoComplete(Autocomplete);

Autocomplete.propTypes = {
  hits: PropTypes.any,
};

Autocomplete.defaultProps = {
  hits: [],
};

class HeaderSearch extends Component {
  state = {
    recommendation: true,
  }

  componentDidMount() {
    window.addEventListener('click', (e) => {
      if (e.target.className !== 'ais-SearchBox-input') {
        this.setState({ recommendation: false });
      }
    });
  }

  onSubmit = (e) => {
    e.preventDefault();
    const { history, searchState } = this.props;
    this.setState({
      recommendation: true,
    });
    history.push(`/?search=${searchState.query || ''}`);
  }

  onKeyDown = () => {
    const { searchState } = this.props;
    this.setState({
      recommendation: searchState.query !== '',
    });
  }

  render() {
    const { recommendation } = this.state;
    const { searchState, _setSearchState } = this.props;
    return (
      <InstantSearch
        indexName={process.env.REACT_APP_ALGOLIA_INDEX_NAME}
        searchClient={searchClient}
        searchState={searchState}
        onSearchStateChange={_setSearchState}
      >
        <SearchBox
          onSubmit={this.onSubmit}
          onKeyDown={this.onKeyDown}
        />
        { searchState.query && recommendation ? <CustomAutocomplete /> : null }
      </InstantSearch>
    );
  }
}

HeaderSearch.propTypes = {
  searchState: PropTypes.object,
  history: PropTypes.object,
  _setSearchState: PropTypes.func,
};

HeaderSearch.defaultProps = {
  searchState: {},
  history: {},
  _setSearchState: () => '',
};

const mapStateToProps = ({ search: { searchState } }) => ({
  searchState,
});

const mapDispatchToProps = (dispatch) => ({
  _setSearchState: (payload) => dispatch(setSearchState(payload)),
});

export default connect(mapStateToProps, mapDispatchToProps)(HeaderSearch);
