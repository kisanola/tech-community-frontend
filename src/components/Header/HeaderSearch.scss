@import '../../styles/colors';

.ais-SearchBox-form {
  background-color: $surface;
  display: flex;
  align-items: center;
  padding: 0.2rem 1rem;
  border-radius: 1.5rem;
  -webkit-border-radius: 1.5rem;
  -moz-border-radius: 1.5rem;
  -ms-border-radius: 1.5rem;
  -o-border-radius: 1.5rem;
  min-width: 400px;
  max-width: 100%;
  input {
    outline: none;
    border: 0;
    background-color: transparent;
    flex: 1;
    height: 2rem;
  }
  .ais-SearchBox-submit,
  .ais-SearchBox-reset {
    padding-right: 0.3rem;
    outline: none;
    background: none;
    border: none;
  }
}

.ais-InstantSearch__root {
  position: relative;
  width: 30rem;
  margin-right: 2rem;
  .search-autocomplete {
    position: absolute;
    top: 55px;
    left: 0;
    right: 0;
    background-color: #fff;
    max-height: 400px;
    overflow-y: hidden;
    height: auto;
    list-style: none;
    padding: 0px;
    li {
      border-bottom: 1px solid $secondary;
      padding: 5px 10px;
      a {
        text-decoration: none;
      }
      &:hover {
        background-color: $surface;
        cursor: pointer;
      }
      p {
        padding: 0;
        margin: 0;
      }
      span {
        background-color: $secondary;
        color: #fff;
        display: inline-block;
        padding: 0 5px;
        border-radius: 2px;
        font-size: 0.9em;
        i {
          margin-right: 5px;
          font-size: 0.7em;
        }
      }
    }
  }
}
