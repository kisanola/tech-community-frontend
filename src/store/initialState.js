module.exports = {
  currentUser: {
    isAuth: localStorage.getItem('token') !== null,
    user: {},
  },
};
