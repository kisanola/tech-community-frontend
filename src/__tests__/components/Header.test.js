import React from 'react';
import { shallow, mount } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';
import { Header, mapStateToProps } from '../../components/Header/Header';
import initialState from '../../store/initialState';

let wrapper;
const props = {
  match: { path: '' },
};
describe('Header.jsx', () => {
  beforeEach(() => {
    wrapper = mount(
      <Router>
        <Header {...props} />
      </Router>,
    );
  });
  test('should render Header.jx', () => {
    wrapper = shallow(<Header {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should render Header.jx', () => {
    const component = wrapper.find('Header');
    expect(component.props().match).toHaveProperty('path');
  });

  test('should render Header.jx with login user', () => {
    const newProps = { ...props, isAuth: true };
    wrapper = mount(
      <Router>
        <Header {...newProps} />
      </Router>,
    );
    const component = wrapper.find('Header');
    expect(component.props().isAuth).toBeTruthy();
  });

  describe('when clicking on `menu` button', () => {
    beforeEach(() => {
      wrapper = mount(
        <Router>
          <Header {...props} />
        </Router>,
      );
    });
    test('should toggle `menu` state', () => {
      wrapper.find('button.navbar-toggler').simulate('click');
      const component = wrapper.find('Header');
      expect(component.state().menu).toBeTruthy();
    });
  });

  describe('when props changes', () => {
    beforeEach(() => {});

    test('should render `/login` as `active', () => {
      props.match.path = '/login';
      wrapper = mount(
        <Router>
          <Header {...props} />
        </Router>,
      );
      const path = '/login';
      const component = wrapper.find('Header');
      expect(component.props().match.path).toBe(path);
    });

    test('should render `/signup` as `active`', () => {
      props.match.path = '/signup';
      wrapper = mount(
        <Router>
          <Header {...props} />
        </Router>,
      );
      const path = '/signup';
      const component = wrapper.find('Header');
      expect(component.props().match.path).toBe(path);
    });
  });

  describe('reducers', () => {
    test('should return `mapStateToProps`', () => {
      const expectedState = {
        isAuth: false,
        user: {},
      };
      const state = mapStateToProps(initialState);
      expect(state).toEqual(expectedState);
    });
  });
});
