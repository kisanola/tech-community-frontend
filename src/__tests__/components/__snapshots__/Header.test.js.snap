// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header.jsx should render Header.jx 1`] = `ShallowWrapper {}`;
